### DISCUSSION TITLE
Update the CSS template for changing the code-block layout

### DISCUSSION THREAD

**Discussion by kristosh** [alshedivat/al-folio#2348] [Q&A] [answered]
I am trying to modify the CSS of the template for changing the code-block layout. I would like to use something similar to sublime CSS just for the code-blocks in the al-folio template. However, I am not really sure how to modify just the specific part of the template. 

Could you please indicate what I need to change in order to update the template in the desired way?

**Comment by george-gca** [comment]
As you can see in our [_config.yml](https://github.com/alshedivat/al-folio/blob/363f277aa82df6e03f172dbcf2f9c1fbe84aca26/_config.yml#L198) we use [Rouge](https://github.com/rouge-ruby/rouge) for syntax highlighting.

I believe the easiest way (if you are going to do a lot of modifications) is for you to get a [readily implemented css class](https://jwarby.github.io/jekyll-pygments-themes/languages/ruby.html) for this, test it, and then add your modifications on top of it. Note that `Rouge` HTML output is compatible with [stylesheets designed for Pygments](https://pygments.org/styles/).

To set up one of these styles simply copy the CSS file into `_sass/` and import the syntax highlighter styles into `_sass/_main.scss`, as stated in [jekyll docs](https://jekyllrb.com/docs/liquid/tags/#stylesheets-for-syntax-highlighting).

**Reply by kristosh** [reply]
Hi George,

I have already created the desired `css` and I have placed it in the `_sass` folder. Not sure if there is another or better place to add it. It seems also that I cannot locate any` _main.scss` file in the folder. Is that normal? I cannot only find the `main.scss` in the `assets/css` folder.

Actually, what I ve tried to do is to create the file using the following command:

`rougify style monokai.sublime > code-block.css
`

Then, I am trying to import this `css` file using the main.scss file. However, it seems that it does not like the system of the `css `file!

**Reply by george-gca** [reply]
Sorry, I meant [assets/css/main.scss](https://github.com/alshedivat/al-folio/blob/master/assets/css/main.scss).

> However, it seems that it does not like the system of the css file!

What do you mean?

**Reply by kristosh** [reply]
Ok so in the main.scss I can see the import and I guess it imports all the files from the _sass that are `scss` and not `css`. When I am trying to upload the file `code-block.css ` I got the following error:

```
1 │ ��.highlight table td { padding: 5px; }
  │ ^
  ╵
  syntax.css 1:1                                                                                     @import
  ...\assets\css\main.scss 11:3  root stylesheet
  Conversion error: Jekyll::Converters::Scss encountered an error while converting 'assets/css/main.scss':
                    Invalid UTF-8
```

I have tried to generate the file using `rougify style monokai.sublime > code-block.scss ` and havent work as well. I guess the generated file has wrong syntax but not sure how to fix it.

In general I have tried to follow the instructions from [here](https://github.com/rouge-ruby/rouge?tab=readme-ov-file#command-line).

**Reply by kristosh** [reply]
The generated css file by the rougify is the following:

```
.highlight table td { padding: 5px; }
.highlight table pre { margin: 0; }
.highlight .gh {
  color: #999999;
}
.highlight .sr {
  color: #f6aa11;
}
.highlight .go {
  color: #888888;
}
.highlight .gp {
  color: #555555;
}
.highlight .gs {
}
.highlight .gu {
  color: #aaaaaa;
}
.highlight .nb {
  color: #f6aa11;
}
.highlight .cm {
  color: #75715e;
}
.highlight .cp {
  color: #75715e;
}
.highlight .c1 {
  color: #75715e;
}
.highlight .cs {
  color: #75715e;
}
.highlight .c, .highlight .ch, .highlight .cd, .highlight .cpf {
  color: #75715e;
}
.highlight .err {
  color: #960050;
}
...
...
...
```

**Reply by george-gca** [reply]
Your file clearly has some hidden chars that are causing the error, as seen in the error message:

```
1 │ ��.highlight table td { padding: 5px; }
  │ ^
  ╵
```

Delete these characters and try again. You can check for hidden chars with sites like [this one](https://www.soscisurvey.de/tools/view-chars.php).

**Reply by kristosh** [reply]
Hey thanks for the recommendation. Firstly, is the dot before the highlight a legit syntax? Is there also a way to remove all the hidden chars somehow without writing the code again?

**Reply by george-gca** [reply]
Yes, it is part of the [CSS syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/Reference). Probably these chars are only at the beginning of the file, so just deleting them in a normal text editor should be fine. While doing that you'll notice that you will press the delete but nothing will happen, but it is because the chars are invisible for your editor. If there are a lot of hidden chars, there are some sites like the one I posted here or programs that might help you, but I have no idea which ones.

**Reply by kristosh** [reply]
Oh that work indeed. Now I can see that the compilation works without a problem. I am facing a slight problem that not all the color frmo sublime-monokai template is employed. My codeblock background is black (as intended) and most of the code text in white. However, all the text that it supposed to be in color (comments, import etc) is black. When I am refreshing the site I can see the color text for a while but then again it turns and renders black. 

![image](https://github.com/alshedivat/al-folio/assets/24369160/155bcce0-ee45-4514-82cf-7e3126083bb7)

Instead of:

![image](https://github.com/alshedivat/al-folio/assets/24369160/1f7f5b63-fe8f-48a7-94b3-69eaae4addd4)

Is that due to the css of the al-folio template that does not allow the color to portrayed?

**Reply by george-gca** [reply]
This is a little bit trickier to debug. Somewhere in the code the css classes are being overwritten. The easier way to find this is by using [chrome dev tools](https://developer.chrome.com/docs/devtools/css) or [firefox dev tools](https://firefox-source-docs.mozilla.org/devtools-user/). In there you can click in the element and find all the attributes that are set for that element and where are they. If the color setting is with a strike through, this means that somewhere above that value is overwritten. You'll need to find where and maybe force it not to be. I would try first to import your css file as the last import possible in the template, since precedence matters.

**Comment by kristosh** [comment]
Hey I have tried to import my file as last in order or first in  order. Both did not help. Then, after some inspection I have noticed that indeed there is a couple of strike through variables. However, these are not in the codeblock.csss file that it should have been the case, right? There is also some scss file imported by bootstrap that makes the read a bit more complicate for me.

By the way, I was just looking for .highlight variables. I am guessing that the codeblock make use of them explicitly for the styling.

**Reply by kristosh** [reply]
What is weird is that when I refresh the page I can see properly the colors but after a while the disappear. So it is loaded properly at first and then they are overwritten somehow!

**Reply by george-gca** [reply]
Yes, that is what happens. Now it is the tricky part, find out where this happens. But you should not make your script the first one imported, since it will have the least priority.

**Reply by kristosh** [reply]
After some debugging I have noticed the following that during the inspection in Firefox, I can only one two lines of code from my codeblock `SCSS` file. The rest of the code is not activated somehow. I am trying to look what could block that, but I cannot find anything related to `.highlight` in the rest of the code! 

Secondly, what it means when during the refresh you can see the colors in the text but then they disappear? Is that because the overwriting taking place?

**Reply by george-gca** [reply]
There is the include of the css files of jekyll pygments that happens on [_includes/head.liquid](https://github.com/alshedivat/al-folio/blob/c4c066501eda87a526d2b3401d1c0f0a98ce1a5e/_includes/head.liquid#L37). About the disappearing, the css values are being overwritten somewhere, but it could even be your css classes overriding the default values for the classes. Usually when something takes a while to be overriden means that they are being loaded post page loading, so they are being [deferred](https://www.w3schools.com/tags/att_script_defer.asp).

**Comment by kristosh** [comment]
Oh, I cannot find the .`liquid` files. In the _includes folder there are only `html` files with similar names. Does that make sense? I guess I do have an older version of the template.

**Reply by george-gca** [reply]
Yes, you have a very old version. They are basically the same files (the html and liquid ones). They were renamed so `prettier` can better handle them, but this was quite a few months ago.

**Reply by kristosh** [reply]
Is there a chance that the conflict is in file that I cannot control, like a boostrap file? I have tried to disable the code from all things above my highlight code from the codeblock.css file. There are some files that imported online I guess from bootstrap.

**Reply by george-gca** [reply]
Yes and no. It might be some external library that is overriding your variables, but even on that case you can force yours to be the last override, either by putting it to be imported last, or by using something like [!important](https://www.w3schools.com/css/css_important.asp). The best way to check what is overriding your values is checking the values on chrome/firefox dev tools. They show the values that were set, and the values that were supposed to be set but overriden with a strike through.

**Reply by kristosh** [reply]
What worries me is that during the inspection with the dev tools, I can see only 4 lines out of the total 120 lines of the imported file to be there (codeblock.scss). There rest lines are not there and are not activated somehow, even when I set in all the commands the !important feature.  I cannot see any strike through lines from my codeblock.css file.  

Moreover, the command highlight that relates with my external [scss](url) file cannot be found in another file.  This issue is really annoying :)

**Reply by kristosh** [reply]
There are also some files like boostrap.css rfs.scss reboot.css that I cannot locate locally in my repository.

**Reply by george-gca** [reply]
The dev tools only displays the css classes that affect the current selected element, that is why you are seeing only a part of the files.

I don't know exactly the internal works of this, but your `scss` files are processed and converted into regular `css` files by [jekyll](https://jekyllrb.com/docs/configuration/sass/). If you import your file inside [assets/css/main.scss](https://github.com/alshedivat/al-folio/blob/master/assets/css/main.scss), it will appear as part of `assets/css/main.css?hash` in the dev tools.

As for `boostrap.css` and some others, these are included in [_includes/head.liquid](https://github.com/alshedivat/al-folio/blob/master/_includes/head.liquid) and loaded via some [CDN](https://www.cloudflare.com/learning/cdn/what-is-a-cdn/). Some of them are only loaded if set in the page (for example, bootstrap table, which has a condition `{% if page.pretty_table %}` in this file), others are loaded for every page (like bootstrap).

**Comment by kristosh** [comment]
I have also noticed that in the updated version of the template there is the following file called `https://github.com/alshedivat/al-folio/blob/master/assets/css/jekyll-pygments-themes-github.cs`s that contains some code related to codeblocks. I was trying to locate whether in the old template there is something similar that causes the overwriting. Are you maybe familiar with these files?

Maybe there is a similar file on the old template as well.

**Reply by george-gca** [reply]
This file was previously included in [_includes/head.liquid](https://github.com/alshedivat/al-folio/blob/master/_includes/head.liquid) (probably `head.html` in your case) and loaded via [CDN](https://www.cloudflare.com/learning/cdn/what-is-a-cdn/). This changed in #1562.

**Reply by kristosh** [reply]
Is there maybe an alternative way to give importance to these commands? Adding the file as a last import in `main.css` and adding everywhere the `important!` line did not work.

**Reply by kristosh** [reply]
Btw, I have tried the same steps in the updated version of al-folio. I have tried to import the `.css` from the `main.scss` file and also add an include in the `head.liquid`. I got exactly the same problem.

**Comment by kristosh** [comment, marked as answer]
After some extra investigation, I figured out that the file responsible for the update of the code is the assets`\js\distillpub\template.v2.js` which is a rather big document. I have just commented out completely. The site looks weird but the `codeblock `has the desired `css`. I guess there is part of this code that blocks that `CSS`.

The js function responsible for that is the _`highlightElement: function(element, async, callback)`_ which after commenting it out, my codeblock works smoothly with the `submline` style. Not sure quite what the implications are if I fully comment on this function in this .`js`  though :)!

**Comment by george-gca** [comment]
Note that the only place where `assets/js/distillpub/template.v2.js` is used is in [_layouts/distill.liquid](https://github.com/alshedivat/al-folio/blob/06ca08cbaf7dbdbbe94c271df4d565c94c8b537f/_layouts/distill.liquid#L18), which in turn is only used if you set `layout: distill` in the page front matter. I had no idea that was your case.

**Comment by kristosh** [comment]
No worries George, thanks for the help! I did not know that this layout will be that crucial. Anyway, I learned a lot investigating this issue. You were constantly helping me finding the issue. I learned how to use the inspections tools from Chrome/Firefox, though I think it would have been impossible to locate the issue from the inspection alone.
